### ISSUE TITLE
Tooltips for messages with an explicit loc broken

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#17023]
The changes made in #16978 don't display tooltips correctly when the warning has a `loc` value.  (I'm not sure that PR was intended to fix this case.)

I added a command that generates a warning with a `loc`.  ~The first time you run it,~ If you run it with 2 `Forward`s the tooltip doesn't show up correctly:

![image](https://user-images.githubusercontent.com/1253341/209490467-ec6bb5ed-2267-4376-bb11-51ab3e18b460.png)

~Resetting Coq and rerunning the script in the same buffer~, If you ~undo and rerun `Foo`~ run both the `Goal` and `Foo` with a single `Run to End`, the tooltip doesn't appear at all.  (You also get this with 2 `Forward`s if `Foo` is not the last line in the buffer.):

![image](https://user-images.githubusercontent.com/1253341/209490542-07772449-80ca-4e29-964a-e15a3ed021b7.png)

My test command:

```
      | IDENT "Foo"; x = string -> {
                                        Printf.eprintf "Some bp ep %d %d \n%!" loc.Loc.bp loc.ep;
                                        let f = CWarnings.create ~name:"Foo" ~category:"scope"
                                            (fun () ->
                                             Pp.strbrk "Foo message") in
                                             f ~loc ();
                                             VernacAbort }
```

Attn: @ejgallego

**Comment by Zimmi48** [comment]
IIUC, this is not a regression. If that's indeed the case, there is no reason to aim to address this in 8.17.

**Comment by ejgallego** [comment]
@jfehrle , you are using the `CWarnings` API incorrectly, instead of emitting a warning, you a creating a new one each time the command is called. As I noted in the other PR, that misunderstanding led you to think that no warning calls `loc`, this is not the case (quite a few warnings set `loc`), however, some other bugs got in the way.

Along with the above, the bug for replay also could have gotten you weird results here.

**Comment by jfehrle** [comment]
> IIUC, this is not a regression.

True, though the code for this case was updated and still doesn't work.  If an easy fix emerges soon, perhaps we can include it and give tooltips a clean bill of health for 8.17.

> you are using the `CWarnings` API incorrectly, instead of emitting a warning, you a creating a new one each time the command is called.

My test code creates and emits a warning each time it's called.  I think recreating it each time doesn't matter.

> the bug for replay also could have gotten you weird results here.

I was looking at this with your new code today and still saw the issue.  I made a little progress, I'll let you know if I find the cause and a simple fix.

**Comment by jfehrle** [comment]
This works as it should if the warning is generated in `idtac`.  Not sure why it doesn't work for `Foo`, which does an `Abort`.

**Comment by ejgallego** [comment]
> Not sure why it doesn't work for `Foo`, which does an `Abort`.

Likely because (as I've pointed out a few times) `attach_tooltip` will never be called in the first place.

**Comment by jfehrle** [comment]
That's not correct.  `attach_tooltip` is called the first time (shown in the first screen shot) but not subsequently (in the second screen shot).  Resetting Coq goes back to the first screen shot result.

**Comment by Zimmi48** [comment]
I can reproduce the issue reported here with existing commands:

```coq
Goal True.
Focus.
```

- in Coq master, the displayed warning does not seem to be associated to a location (no underlined text and thus no tooltip),
- however, in Coq 8.15-8.16, the displayed warning is associated to a (correctly located) underline, but the tooltip is broken, and if we undo and redo the last line, it does not show at all (exactly as described for the `Foo` command in this issue),
- in Coq 8.8-8.14, the tooltip was working, but the location was wrong.

**Comment by jfehrle** [comment]
Can you think of a test case on master where the warning is on a tactic?  Perhaps a deprecation?

When I examined the behavior for `Foo` and `idtac`, it appeared that Coq was returning the correct stateid in both cases.  In master, the first few lines of `Coqops.process_feedback_msg` try to lookup the sentence in the document.  If that fails, `sentence` is None and the match goes to case that doesn't add a tooltip at all.  But the root cause, I dunno.

Also BTW, I noticed that `Doc.find_map` is O(n*n) where n is the number of sentences.  That may be worth improving; I earlier noticed that running the large script `Pff.v` (28K lines) used more than a minute of CPU time for CoqIDE alone.

**Comment by jfehrle** [comment]
I noticed another thing: I described the behavior when I use two `Forward` operations to run the script.  If I instead use `Run to End`, I never see an underline and thus no tooltip, either.

**Comment by ejgallego** [comment]
What I mean is that it would be very surprising that the code in `attach_tooltip` that I fixed is bad; the issue(s) are elsewhere. In particular:

- quite a few times `attach_tooltip` wont' be called due to the sentence not having a `Stateid.t`
- in the case @Zimmi48 has provided with focus, the location sent to `attach_tooltip` is incorrect, in particular it is
```
line: -1, col: 11 -- line: -1, col: 16 / {11-16}
```
So with `line: -1` no wonder the tooltip is not attached properly. Coq core sends the location correctly (coq-lsp has no problems with this case), so the location is mangled somewhere in the xmlprotocol / coqide stack.

**Comment by jfehrle** [comment]
The XML protocol only sends the start and stop positions for `loc`.  Changing the XML protocol to add the line number would disrupt the other GUIs using the protocol.  Surely there's better way to address this.

For example:

```
<feedback object="state" route="0">
  <state_id val="2"/>
  <feedback_content val="message">
    <message>
      <message_level val="warning"/>
      <option val="some">
        <loc start="12" stop="17"/>
      </option>
```

**Comment by jfehrle** [comment]
The other problem is on the Coq side.  The first lines below are for my "idtac" message, for which the tooltip appears correctly.  The second group is for "Focus", which fails.  I expect that's because the ASSIGN_TIP is received _after_ ATTACH_TOOLTIP.

PUSH
**ASSIGN_TIP 10**
idtac Some bp ep 12 21
**ATTACH_TOOLTIP start = 12 stop = 21**

PUSH
**ATTACH_TOOLTIP start = 25 stop = 30
ASSIGN_TIP 11**

The order is different because tactics are not evaluated until the subsequent `SubGoals` XML call invokes `Stm.finish`, while commands are processed immediately.  This bug may date back to b2f27276 when `Stm` was created.  One way to fix it would be to add a new message to send the stateid before the command/tactic is executed, outside of the request/reply mechanism (like feedback messages).

**Comment by ejgallego** [comment]
> The XML protocol only sends the start and stop positions for loc.

Oh wow! I had forgotten about this, things are quite broken then.

In this sense, the only fix that makes sense for Coq is to fix this and stop sending bogus locations.

- scanning from the beginning of buffer: will incur in performance problems
- the cache you propose: there is nothing that guarantees order
- doing proper caching (based on line) is what GTK already implements, would require a rewrite 

As for the tip problem, I bet the problem requires a protocol re-design to fix (as we did back in 2016 with SerAPI) , adding more synchronous messages will only make matters worse as the size number of transitions in the automaton increases non-linearly.

**Comment by jfehrle** [comment]
> the only fix that makes sense for Coq is to fix this and stop sending bogus locations

The locations are incomplete, not bogus (ie the information sent is correct).  Changing what's included in the messages for `loc` likely requires code changes to vsCoq and Coqtail and may be so simple for them.  Let's not jump to the impactful "perfect" solution too quickly.

> scanning from the beginning of buffer: will incur in performance problems

It depends on how many messages with `loc` the script generates and the length of the script.  Both of those need to be large to be noticeable.  The main performance issue addressed in 8.15 was for highlighting each sentence as it is processed--a different case.  IIRC there were a couple hundred warning messages in `Pff.v` in 8.15 from a newly-added message but the extra time required for the conversion on that file was fairly small, less than a doubling.  And this goes away when you fix the messages, unlike the highlighting that always has to happen.

> the cache you propose: there is nothing that guarantees order

It's not an absolute order, so if the location to convert is before the saved point then you examine items from the beginning of the buffer.  The really effective cache would be to save the starting byte offset for the sentence as I did in the previous code (though the code was buggy).

> As for the tip problem, I bet the problem requires a protocol re-design to fix

I believe I can work around this fairly simply without changing the protocol.

**Comment by jfehrle** [comment]
There's another case to consider.  I run this script with "Run to End":

![image](https://user-images.githubusercontent.com/1253341/209992296-b0fba063-fd5c-4066-99bc-d8f767b1368d.png)

If I then add a few spaces before `slow` completes, the tooltip highlight is mispositioned.  The tooltip displays correctly if the cursor is on the `idtac`:

![image](https://user-images.githubusercontent.com/1253341/209992819-d3a974ae-1fd8-409a-9b75-0393699b4be8.png)

If the cursor is not over `idtac` but over the `w.` of `slow`, the tooltip doesn't display correctly:

![image](https://user-images.githubusercontent.com/1253341/209993094-7dcc5c72-3c9b-40a0-9950-63ef1d767b50.png)

This can be fixed by saving the original buffer offset of the sentence that's sent in the `add` request.  Then when a message with an offset is received, apply an adjustment (the difference between the current beginning of sentence position from the mark and the offset sent in `add`).

**Comment by jfehrle** [comment]
I have a single-line fix for empty tooltips on commands (e.g. `Focus`).

@ejgallego The fix for the problem in the comment just above this one requires (re)introducing a field or two in the `sentence` record, also changing `coq_loc_to_gtk_offset`.  Not sure if `coq_loc_to_gtk_offset` can still use `byte_off_to_buffer_off`.  This approach would also avoid the need to pass `line_nb` in the protocol.  If that's agreeable, I can provide code.  Or do it yourself.  Let me know.

**Comment by jfehrle** [comment]
> @Zimmi48 IIUC, this is not a regression.

Actually, it is.  Since the protocol doesn't send the lineno as part of the loc, the changes in #16978 fail to highlight _any_ tooltips for, say, `Pff.v`, which has many warnings.  We should fix this for 8.17.  I'm working on this.

**Comment by Zimmi48** [comment]
This is fixed in v8.17, but I'm leaving the issue open because this is not yet fixed in master (several options are available and one of them should be merged).

**Comment by ejgallego** [comment]
This was fixed by #17382
